File: client/src/main/scala/stasis/client/ops/recovery/stages/EntityCollection.scala
package stasis.client.ops.recovery.stages

import org.apache.pekko.NotUsed
import org.apache.pekko.stream.Materializer
import org.apache.pekko.stream.scaladsl.Source

import stasis.client.model.DatasetMetadata
import stasis.client.model.FilesystemMetadata
import stasis.client.model.TargetEntity
import stasis.client.ops.Metrics
import stasis.client.ops.ParallelismConfig
import stasis.client.ops.recovery.Providers
import stasis.shared.ops.Operation

trait EntityCollection {
  protected def targetMetadata: DatasetMetadata
  protected def keep: (String, FilesystemMetadata.EntityState) => Boolean
  protected def destination: TargetEntity.Destination
  protected def providers: Providers
  protected def parallelism: ParallelismConfig

  protected implicit def mat: Materializer

  private val metrics = providers.telemetry.metrics[Metrics.RecoveryOperation]

  def entityCollection(implicit operation: Operation.Id): Source[TargetEntity, NotUsed] =
    Source(providers.kinds.toList)
      .flatMapConcat(_.collector(targetMetadata, keep, destination, providers, parallelism).collect(providers.filesystem))
      .wireTap { entity =>
        metrics.recordEntityExamined(entity = entity)

        providers.track.entityExamined(
          entity = entity.ref,
          metadataChanged = entity.hasChanged,
          contentChanged = entity.hasContentChanged
        )
      }
      .filter(_.hasChanged)
      .wireTap { entity =>
        metrics.recordEntityCollected(entity = entity)
        providers.track.entityCollected(entity = entity)
      }
}

File: client-android/lib/src/main/kotlin/stasis/client_android/lib/ops/recovery/stages/EntityCollection.kt
package stasis.client_android.lib.ops.recovery.stages

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.emitAll
import kotlinx.coroutines.flow.filter
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.onEach
import stasis.client_android.lib.model.DatasetMetadata
import stasis.client_android.lib.model.FilesystemMetadata
import stasis.client_android.lib.model.TargetEntity
import stasis.client_android.lib.ops.OperationId
import stasis.client_android.lib.ops.recovery.Providers
import java.nio.file.FileSystem

interface EntityCollection {
    val targetMetadata: DatasetMetadata
    val keep: (String, FilesystemMetadata.EntityState) -> Boolean
    val destination: TargetEntity.Destination
    val providers: Providers

    fun entityCollection(operation: OperationId, filesystem: FileSystem): Flow<TargetEntity> =
        flow {
            providers.kinds.forEach { kind ->
                emitAll(kind.collector(targetMetadata, keep, destination, providers).collect(filesystem))
            }
        }
            .onEach { entity ->
                providers.track.entityExamined(
                    operation = operation,
                    entity = entity.ref,
                    metadataChanged = entity.hasChanged,
                    contentChanged = entity.hasContentChanged
                )
            }
            .filter { it.hasChanged }
            .onEach { entity ->
                providers.track.entityCollected(operation = operation, entity = entity)
            }
}
